File: packages/app/src/services/software/definitions/streamlink.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory } from 'shared';
import { GithubReleasesScraper } from '../githubReleasesScraper';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'streamlink';
	public name = 'Streamlink';
	public category = [SoftwareCategory.Media];
	public downloadName = 'streamlink-x86_64.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public deprecated = false;
	public icon = 'streamlink.png';
	public homepage = 'https://streamlink.github.io';

	public async resolveDownloadUrl() {
		const owner   = 'streamlink';
		const repo    = 'windows-builds';
		const release = await GithubReleasesScraper.getRelease(owner, repo, r => r.match(/streamlink-\d+\.\d+\.\d+(?:-\d+)?-py3\d+-x86_64\.exe/) !== null);
		if (release.isErr()) {
			return err(release.error);
		}

		return ok(release.value);
	}
}

File: src/lib/software/streamlink.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from src.lib.github_release_scraper import GithubReleaseScraper

class Streamlink(BaseSoftware):
    def __init__(self):
        super().__init__()

        self._gh = GithubReleaseScraper('streamlink', 'windows-builds', self)
        self._gh.releases_scraped.connect(self._on_releases_scraped)

        self.key = 'streamlink'
        self.name = 'Streamlink'
        self.category = [SoftwareCategory.Media]
        self.download_name = 'streamlink-x86_64.exe'
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'streamlink.png'
        self.homepage = 'https://streamlink.github.io'

    @Slot(list)
    def _on_releases_scraped(self, releases: list[str]):
        asset_pattern = compile(r'streamlink-\d+\.\d+\.\d+(?:-\d+)?-py3\d+-x86_64\.exe')
        asset = next((release for release in releases if asset_pattern.search(release)), None)
        if asset:
            self.url_resolved.emit(asset)
        else:
            self.url_resolve_error.emit(self.ResolveError.GitHubAssetNotFoundError)

    def resolve_download_url(self):
        self._gh.get_repo_releases()
